### ISSUE TITLE
Race condition during fetching

### ISSUE THREAD

**Issue by adamjstewart** [spack/spack#10096]
I encountered a race condition during package fetching. In two separate terminals, I tried installing two packages that both happened to depend on Perl. Both processes crashed at exactly the same time with slightly different error messages:
```console
==> No binary for perl found: installing from source
==> Fetching http://www.cpan.org/src/5.0/perl-5.26.2.tar.gz
######################################################################## 100.0%
==> Fetching http://search.cpan.org/CPAN/authors/id/M/MI/MIYAGAWA/App-cpanminus-1.7042.tar.gz
######################################################################## 100.0%
==> Error: ChecksumError: md5 checksum failed for /Users/Adam/spack/var/spack/stage/perl-5.26.2-uhov4vq4lk3u7biruczmxbytm22agi66/perl-5.26.2.tar.gz
    Expected dc0fea097f3992a8cd53f8ac0810d523 but got 561f783354f933c81d1c852950f70426

/Users/Adam/spack/lib/spack/spack/package.py:920, in do_fetch:
        917        self.stage.fetch(mirror_only)
        918        self._fetch_time = time.time() - start_time
        919
  >>    920        if checksum and self.version in self.versions:
        921            self.stage.check()
        922
        923        self.stage.cache_local()
```
```console
==> No binary for perl found: installing from source
==> Fetching http://www.cpan.org/src/5.0/perl-5.26.2.tar.gz
######################################################################## 100.0%
==> Error: FileNotFoundError: [Errno 2] No such file or directory: '/Users/Adam/spack/var/spack/stage/perl-5.26.2-uhov4vq4lk3u7biruczmxbytm22agi66/perl-5.26.2.tar.gz.part' -> '/Users/Adam/spack/var/spack/stage/perl-5.26.2-uhov4vq4lk3u7biruczmxbytm22agi66/perl-5.26.2.tar.gz'

/Users/Adam/spack/lib/spack/spack/package.py:916, in do_fetch:
        913                raise FetchError("Will not fetch %s" %
        914                                 self.spec.format('$_$@'), ck_msg)
        915
  >>    916        self.stage.create()
        917        self.stage.fetch(mirror_only)
        918        self._fetch_time = time.time() - start_time
        919
```
When I deleted the partial download and started over, things worked fine. I think we need to put a lock on the download cache directory during fetching.

@tgamblin

**Comment by adamjstewart** [comment]
I also noticed a race condition during installation. When two processes install the same package at the same time, one of them notices that the installation prefix exists, but no `.spack` subdirectory exists, so you get an error like:
```
==> Error: [Errno 2] No such file or directory: '/mnt/a/u/sciteam/stewart1/spack/opt/spack/cray-cnl5-interlagos/gcc-7.3.0/cairo-1.16.0-ss55jg7anitg2apd62hjdrjsf3r3bi7w/.spack'
```
In the meantime, the other process successfully installs the package, but then hits:
```
==> Error: OSError: [Errno 2] No such file or directory: '/mnt/a/u/sciteam/stewart1/spack/var/spack/stage/cairo-1.16.0-ss55jg7anitg2apd62hjdrjsf3r3bi7w'

/mnt/a/u/sciteam/stewart1/spack/lib/spack/spack/package.py:1248, in _stage_and_write_lock:
       1245    def _stage_and_write_lock(self):
       1246        """Prefix lock nested in a stage."""
       1247        with self.stage:
  >>   1248            with spack.store.db.prefix_write_lock(self.spec):
       1249                yield
```
Not sure if the first process decides to delete the directory since it thinks it is corrupted or what.

**Comment by tgamblin** [comment]
This is fixed by #13100.
